### ISSUE TITLE
I think your transformer implementation is flawed

### ISSUE THREAD

**Issue by Sea-Snell** [pytorch/pytorch#55270]
Hi, I've been doing some benchmarking with various implementations of transformers on language modeling, and I noticed that the specifics of your transformer implementation/initialization make it unstable when training deeper 12+ layer transformers.

There are two main variants of transformers, post and pre norm (this basically just has to do with where the layer norms are placed, but it's an important distinction for training stability, see [this presentation](https://tnq177.github.io/data/transformers_without_tears.pdf)).

You guys use post norm. This is ok, but it doesn't work on deeper transformers with the specific Xavier-uniform initialization you use. Let me explain:

Both pre- and post-norm perform similarly when tuned. Pre-norm usually works with most generic initializations and doesn't usually require LR-warmup to get decent performance. Post-norm, on the other hand, is usually hard to train for deeper networks (12+ layers), it requires precise initialization and a good LR schedule; this is because putting the layer norm after the residual causes each layer's Jacobian singular values to not be closely distributed around 1, which causes vanishing gradients (see figure 5 of [this paper](https://arxiv.org/pdf/2003.04887.pdf)). With the right initialization, this issue can be overcome, but the specific initialization you guys use does not fix this at all.

In fact for deeper transformers, using the kaiming-uniform initialization (this is the default for Linear layers) is much better than the custom Xavier-uniform initialization you are using.

To prove this I've run some experiments. If you train a language model with the GPT-3-small architecture hyper-parameters (this has 12 transformer layers) on wiki-text-103, the validation cross-entropy will basically flatline at about 7 after approx. one-hundred iterations of training. On the other hand if you commit out line 66 of `pytorch/torch/nn/modules/transformer.py` – `self._reset_parameters()` – the model will actually train for several epochs before converging at around 3.6 validation cross-entropy.

I've attached plots of the training and validation loss for these experiments below (NOTE: the runs labeled with pytorch are from your transformer implementation, the ones labeled with custom are from my own implementation; I cut some of the runs short because I didn't have time, but I let one run from my own custom implementation run for much longer, the pytorch ones track my own very closely, so they should converge similarly. Xavier-uniform is what your implementation is currently set to, kaiming-uniform is the initialization you get from commenting out just one line of your code)

![W B Chart 4_2_2021, 9_11_36 PM](https://user-images.githubusercontent.com/6655321/113467570-97082d00-93f8-11eb-88fd-f2149acde745.png)
![W B Chart 4_2_2021, 9_11_28 PM](https://user-images.githubusercontent.com/6655321/113467573-9cfe0e00-93f8-11eb-9a12-37a40928b617.png)

To fix this there are two things you could do: either change your current default initialization scheme or switch to a pre-norm transformer. I recommend the pre-norm route since its more stable to train in general. Or you could make it an input parameter, so that users can choose which version (pre- or post- norm) they want to use. I'd actually be down to make this fix; I can make the changes and then submit a pull request, I just wanted to see what you think first. What was your thinking when implementing this ? It looks like you just went off the Attention Is All You Need paper, which is maybe a little old now.

Also just in general, I feel like there could be better abstractions for transformers; most researchers don't even use these transformer building blocks you have, instead they just use their own implementation or something else (this is in contrast to your conv-net building blocks, which are easy to work with and pretty flexible and widely used. Maybe its just cause transformers haven't had as much time to mature as conv-nets idk). I don't think any framework gets this right though. Anyways, this is a much bigger question, idk if you want to look into this or if you are already working on this, but I thought I would bring it up.

(P.S. one more thing, because you use post-norm, I think the extra norm you apply after all the layers in the TransformerEncoder/Decoder modules is redundant; this would be necessary if you were using pre-norm though)

cc @zhangguanheng66

**Comment by ezyang** [comment]
cc @jbschlosser as well

**Comment by jbschlosser** [comment]
Thanks @Sea-Snell for the empirically-backed report!

> Also just in general, I feel like there could be better abstractions for transformers; most researchers don't even use these ? transformer building blocks you have, instead they just use their own implementation or something else (this is in contrast to your conv-net building blocks, which are easy to work with and pretty flexible and widely used. Maybe its just cause transformers haven't had as much time to mature as conv-nets idk).

Certainly agree that the transformer abstractions could be better! This is a known problem and we have active, long-term efforts underway to reimplement transformer-related building blocks in such a way that they are both more flexible and performant. I believe the initial thinking was indeed to match Attention Is All You Need, and it became quickly clear that the Transformer modules as originally designed are too rigid to keep up with the research.

> To fix this there are two things you could do: either change your current default initialization scheme or switch to a pre-norm transformer. I recommend the pre-norm route since its more stable to train in general. Or you could make it an input parameter, so that users can choose which version (pre- or post- norm) they want to use. I'd actually be down to make this fix; I can make the changes and then submit a pull request, I just wanted to see what you think first.

In the short term (before the larger refactoring lands), I think providing the option to choose between pre- and post- layer norm is useful. I'd be willing to review this addition if it's done in a backwards-compatible way.

> (P.S. one more thing, because you use post-norm, I think the extra norm you apply after all the layers in the TransformerEncoder/Decoder modules is redundant; this would be necessary if you were using pre-norm though)

There are quite a few issues decrying the redundant layer norm (https://github.com/pytorch/pytorch/issues/24930, https://github.com/pytorch/pytorch/issues/50086, https://github.com/pytorch/pytorch/issues/51447) and I'd love to have this addressed in the short term as well if it can be done in a backwards-compatible way. Feel free to add me as reviewer for any PR you open addressing these things!

**Comment by bhosmer** [comment]
Added pre/post LayerNorm config param in #60593. 

Note: the redundant layer norm issue remains when running LayerNorm-after (the default, for compatibility) - we're leaving that to be addressed in a followup, as there are some BC subtleties around serialized models. But I'm going to go ahead and close this issue, since there are several other issues discussing that problem (see above).
